contributors/design-proposals/customresources-subresources.md

```diff
@@ -0,0 +1,184 @@
+    1. [API Types](#api-types)
+    2. [Feature Gate](#feature-gate)
+5. [Semantics](#semantics)
+    1. [Validation Behavior](#validation-behavior)
+        1. [Status](#status)
+        2. [Scale](#scale)
+    2. [Status Behavior](#status-behavior)
+    3. [Scale Behavior](#scale-behavior)
+        1. [Status Replicas Behavior](#status-replicas-behavior)
+        2. [Selector Behavior](#selector-behavior)
+4. [Implementation Plan](#implementation-plan)
+5. [Alternatives](#alternatives)
+    1. [Scope](#scope)
+
+## Abstract
+
+[CustomResourceDefinitions](https://github.com/kubernetes/community/pull/524) (CRDs) were introduced in 1.7. The objects defined by CRDs are called CustomResources (CRs). Currently, we do not provide subresources for CRs.
+
+However, it is one of the [most requested features](https://github.com/kubernetes/kubernetes/issues/38113) and this proposal seeks to add  `/status` and `/scale` subresources for CustomResources.
+
+## Goals
+
+1. Support status/spec split for CustomResources:
+    1. Status changes are ignored on the main resource endpoint.
+    2. Support a `/status` subresource HTTP path for status changes.
+    3. `metadata.Generation` is increased only on spec changes.
+2. Support a `/scale` subresource for CustomResources.
+3. Maintain backward compatibility by allowing CRDs to opt-in to enable subresources.
+4. If a CustomResource is already structured using spec/status, allow it to easily transition to use the `/status` and `/scale` endpoint.
+5. Work seamlessly with [JSON Schema validation](https://github.com/kubernetes/community/pull/708).
+
+## Non-Goals
+
+1. Allow defining arbitrary subresources i.e. subresources except `/status` and `/scale`.
+
+## Proposed Extension of CustomResourceDefinition
+
+### API Types
+
+The addition of the following external types in `apiextensions.k8s.io/v1beta1` is proposed:
+
+```go
+type CustomResourceDefinitionSpec struct {
+    ...
+    // SubResources describes the subresources for CustomResources
+    // This field is alpha-level and should only be sent to servers that enable
+    // subresources via the CurstomResourceSubResources feature gate.
+    // +optional
+    SubResources *CustomResourceSubResources `json:"subResources,omitempty"`
+}
+
+// CustomResourceSubResources defines the status and scale subresources for CustomResources.
+type CustomResourceSubResources struct {
+    // Status denotes the status subresource for CustomResources
+    Status *CustomResourceSubResourceStatus `json:"status,omitempty"`
+    // Scale denotes the scale subresource for CustomResources
+    Scale  *CustomResourceSubResourceScale  `json:"scale,omitempty"`
+}
+
+// CustomResourceSubResourceStatus defines how to serve the HTTP path <CR Name>/status.
+type CustomResourceSubResourceStatus struct {
+}
+
+// CustomResourceSubResourceScale defines how to serve the HTTP path <CR name>/scale.
+type CustomResourceSubResourceScale struct {
```

Comment:
How are you going to handle versioning of the scale object you're returning?  Right now, there's only one reasonable choice (`Scale.v1.autoscaling`), but to handle a future change that won't break API compatibility for the CustomResource, don't you have to make that explicit?

Reply:
Good point. It's about this mapping server side for the CR scale object: https://github.com/kubernetes/kubernetes/pull/53743/commits/d61a2d90372c301dd11088df8941acf2bb01c38c#diff-49992b38a536f6f81a4100abe29c1334R38

What are our options? We can put a GVK here, defaulting and validating to autoscaling/v1.Scale today. When v2 comes out we allow both and when v1 is deprecated, we deprecate the v1 value here as well. Technically, we don't have to add this GVK fields today, but should do so reasonably early before a v1 deprecation comes to the table. But having a plan and implementing that today doesn't harm either.

We can accept v1 and v2 scale types on the endpoint without a problem. But we can only send one of them to the client and could control that with the GVK field. This would be a hard switch as soon as the CRD is updated. To make this smooth on the (go-)client side one could make the polymorphic discovery client to handle both types gracefully.

@deads2k What are you thoughts about this?

Reply:
> @deads2k What are you thoughts about this?

I agree with your idea of baking it in early. It mirrors my thoughts.